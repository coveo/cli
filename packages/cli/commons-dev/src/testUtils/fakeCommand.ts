import {Command} from '@oclif/core';

export const getFakeCommand = (
  overrideConfig?: Record<string, unknown>
): Command => {
  const fakeCommand = {
    id: 'foo',
    warn: jest.fn(),
    log: jest.fn(),
    parse: jest.fn().mockReturnValue({flags: {}}),
    ...overrideConfig,
  };

  return fakeCommand as unknown as Command;
};
