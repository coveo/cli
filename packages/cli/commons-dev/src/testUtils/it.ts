import {fancy} from 'fancy-test';

/**
 * Wrapper around `jest.it` that captures all stdout/stderr.
 */
export const fancyIt = () => fancy.stdout().stderr().it;
