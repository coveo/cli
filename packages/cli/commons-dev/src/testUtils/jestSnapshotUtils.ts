// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string) =>
  input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/gm, '›');

export const formatAbsolutePath = (input: string) =>
  input.replaceAll(/\w:\\{2}/gm, '/').replaceAll(/\\{2}/gm, '/');

export const removeDotCmd = (input: string) => input.replaceAll(/\.cmd/gm, '');
