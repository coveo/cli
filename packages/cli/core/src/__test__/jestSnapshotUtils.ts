import strip from 'strip-ansi';

// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string, stripAinsi = false) => {
  const formatted = input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/g, '›');
  return stripAinsi ? strip(formatted) : formatted;
};
