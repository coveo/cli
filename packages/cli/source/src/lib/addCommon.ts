import {CliUx} from '@oclif/core';
import {red, dim, green} from 'chalk';
import {BuiltInTransformers, errors} from '@coveo/push-api-client';
import {withNormalizeInvalidFields} from './commonFlags';
import {pluralizeIfNeeded} from '@coveo/cli-commons/utils/string';
import dedent from 'ts-dedent';

const allowedCharRegExp = /^[a-z]+[a-z0-9_]*$/;
interface PrintableUnsupportedField {
  original: string;
  normalized: string;
  valid: boolean;
}

export const formatErrorMessage = (err: unknown) => {
  const {UnsupportedFieldError} = errors;
  if (err instanceof UnsupportedFieldError) {
    const normalizations = getNormalizations(err.unsupportedFields);
    const {fixables, unfixables} = colorizeFields(normalizations);
    const isFixable = unfixables.length === 0;

    printInvalidFieldTable(fixables);
    printInvalidFieldTable(unfixables);
    err.message = getFormattedMessage(isFixable);
  }
};

const getFormattedMessage = (isFixable: boolean) => {
  const normalizeFlag = Object.keys(withNormalizeInvalidFields())[0];
  let message = `Run the same command using the \`--${normalizeFlag}\` flag to automatically normalize field names while pushing your data.`;
  if (!isFixable) {
    message = dedent`Cannot normalize some of the invalid field names detected in your data
      Review your data and ensure all field names as shown in the second table have at least one alphabetic character.`;
  }
  return message;
};

const colorizeFields = (fields: PrintableUnsupportedField[]) => {
  const fixables: PrintableUnsupportedField[] = [];
  const unfixables: PrintableUnsupportedField[] = [];
  for (const field of fields) {
    if (field.valid) {
      const normalized = green(field.normalized);
      fixables.push({...field, normalized});
    } else {
      const normalized =
        field.normalized === ''
          ? dim('(empty field name)')
          : red(field.normalized);
      unfixables.push({...field, normalized});
    }
  }
  return {fixables, unfixables};
};

const getNormalizations = (
  unsupportedFields: [string, string][]
): PrintableUnsupportedField[] => {
  return unsupportedFields.map(([original, transformed]) => {
    const normalized = BuiltInTransformers.toLowerCase(transformed);
    return {
      original,
      normalized,
      valid: isFieldNameValid(normalized),
    };
  });
};

const printInvalidFieldTable = (fields: {valid: boolean}[]) => {
  const count = fields.length;
  const pluralized = pluralizeIfNeeded(['field', 'fields'], count);
  if (count === 0) {
    return;
  }

  const fixable = fields[0].valid;
  CliUx.ux.log(
    ` ${count} ${pluralized} detected in your data can${
      fixable ? '' : 'not'
    } be normalized`
  );
  CliUx.ux.table(fields, {
    original: {header: 'Original'},
    normalized: {header: 'Normalized'},
  });
  logNewLine();
};

export const logNewLine = (lines = 1) => {
  for (let i = 0; i < lines; i++) {
    CliUx.ux.log('');
  }
};

const isFieldNameValid = (fieldName: string): boolean => {
  return allowedCharRegExp.test(fieldName.toLowerCase());
};
