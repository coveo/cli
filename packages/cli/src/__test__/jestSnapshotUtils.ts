// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string) =>
  input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/g, '›');
