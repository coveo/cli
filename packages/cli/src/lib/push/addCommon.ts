import {CliUx} from '@oclif/core';
import {red, dim, green} from 'chalk';
import {BuiltInTransformers, errors} from '@coveo/push-api-client';
import {normalizeInvalidFields} from '../flags/sourceCommonFlags';
import {pluralizeIfNeeded} from '../utils/string';
import dedent from 'ts-dedent';

const allowedCharRegExp = new RegExp('^[a-z]+[a-z0-9_]*$');

export const handleAddError = (err: unknown) => {
  const {UnsupportedFieldError} = errors;
  if (err instanceof UnsupportedFieldError) {
    CliUx.ux.error('Invalid field name detected while parsing your data.', {
      exit: false,
    });
    const normalizations = err.unsupportedFields.map((f) => {
      const original = f[0];
      const normalized = BuiltInTransformers.toLowerCase(f[1]);
      return {
        original,
        normalized,
        valid: isFieldNameValid(normalized),
      };
    });

    const fixable = normalizations
      .filter(({valid}) => valid)
      .map((f) => {
        f.normalized = green(f.normalized);
        return f;
      });

    const unfixable = normalizations
      .filter(({valid}) => !valid)
      .map((f) => {
        const normalized =
          f.normalized === '' ? dim('(empty field name)') : red(f.normalized);
        f.normalized = normalized;
        return f;
      });

    printInvalidFieldTable(fixable);
    printInvalidFieldTable(unfixable);

    const normalizeFlag = Object.keys(normalizeInvalidFields())[0];
    let message = `Run the same command using the --${normalizeFlag} flag to automatically normalize field names while pushing your data.`;
    if (unfixable.length > 0) {
      message = dedent`Cannot normalize some of the invalid field names detected in your data
      Review your data and ensure all field names as shown in the second table have at least one alphabetic character.`;
    }
    logNewLine(2);
    CliUx.ux.error(message, {exit: false});
    logNewLine();
  }
};

const printInvalidFieldTable = (fieldTransformations: {valid: boolean}[]) => {
  const count = fieldTransformations.length;
  const pluralized = pluralizeIfNeeded(['field', 'fields'], count);
  const fixable = fieldTransformations[0].valid;
  if (count === 0) {
    return;
  }

  logNewLine();
  CliUx.ux.log(
    ` ${count} ${pluralized} detected in your data that can${
      fixable ? '' : 'not'
    } be normalized`
  );
  CliUx.ux.table(fieldTransformations, {
    original: {header: 'Original'},
    normalized: {header: 'Normalized'},
  });
};

const logNewLine = (lines = 1) => {
  for (let i = 0; i < lines; i++) {
    CliUx.ux.log('');
  }
};

export const isFieldNameValid = (fieldName: string): boolean => {
  return allowedCharRegExp.test(fieldName.toLowerCase());
};
