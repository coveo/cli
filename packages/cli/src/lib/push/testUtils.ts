export const doMockAxiosError = (
  status: number,
  message: string,
  errorCode: string
) => ({
  response: {
    status,
    data: {
      errorCode,
      message,
    },
  },
});

export const doMockAxiosSuccess = (status: number, text: string) => ({
  status,
  statusText: text,
});
