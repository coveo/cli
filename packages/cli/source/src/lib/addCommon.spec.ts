import {stdout} from 'stdout-stderr';
import {BuiltInTransformers, errors} from '@coveo/push-api-client';
import {formatCliLog} from '@coveo/cli-commons-dev/testUtils/jestSnapshotUtils';
import {formatErrorMessage} from './addCommon';

describe('addCommon', () => {
  const fixableFields: [string, string][] = ['0Invalid-Field', 'poster-A'].map(
    (field) => [field, BuiltInTransformers.identity(field)]
  );
  const unfixableFields: [string, string][] = ['', '$_(*&'].map((field) => [
    field,
    BuiltInTransformers.identity(field),
  ]);

  beforeEach(() => {
    stdout.start();
  });

  afterEach(() => {
    stdout.stop();
  });

  it('should let user know possibility of normalization', () => {
    const error = new errors.UnsupportedFieldError(...fixableFields);
    expect(error.message).not.toContain('Cannot normalize');
  });

  it('should let user know impossibility of normalization', () => {
    const error = new errors.UnsupportedFieldError(...unfixableFields);
    formatErrorMessage(error);
    expect(error.message).toContain('Cannot normalize');
  });

  it('should log fixable fields', () => {
    const error = new errors.UnsupportedFieldError(...fixableFields);
    formatErrorMessage(error);
    expect(formatCliLog(stdout.output)).toMatchSnapshot();
  });

  it('should log unfixable fields', () => {
    const error = new errors.UnsupportedFieldError(...unfixableFields);
    formatErrorMessage(error);
    expect(formatCliLog(stdout.output)).toMatchSnapshot();
  });

  it('should log both fixable and unfixable fields', () => {
    const error = new errors.UnsupportedFieldError(
      ...fixableFields,
      ...unfixableFields
    );
    formatErrorMessage(error);
    expect(formatCliLog(stdout.output)).toMatchSnapshot();
  });
});
