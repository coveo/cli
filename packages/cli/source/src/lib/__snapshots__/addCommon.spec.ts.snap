// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addCommon should log both fixable and unfixable fields 1`] = `
" 2 fields detected in your data can be normalized
 Original       Normalized   
 ────────────── ──────────── 
 0Invalid-Field invalidfield 
 poster-A       postera      

 2 fields detected in your data cannot be normalized
 Original Normalized         
 ──────── ────────────────── 
          (empty field name) 
 $_(*&    (empty field name) 

"
`;

exports[`addCommon should log fixable fields 1`] = `
" 2 fields detected in your data can be normalized
 Original       Normalized   
 ────────────── ──────────── 
 0Invalid-Field invalidfield 
 poster-A       postera      

"
`;

exports[`addCommon should log unfixable fields 1`] = `
" 2 fields detected in your data cannot be normalized
 Original Normalized         
 ──────── ────────────────── 
          (empty field name) 
 $_(*&    (empty field name) 

"
`;
